File: rnamake/motif_topology.py
import collections

import motif_graph
import motif_tree
import motif_type
import motif
import resource_manager as rm
import exceptions


class GraphtoTree(object):

    def __init__(self):
        self.mt = motif_tree.MotifTree()

    class _GraphtoTreeNode(object):
        def __init__(self, parent, parent_end_index, node):
            self.parent = parent
            self.parent_end_index = parent_end_index
            self.node = node
            self.motif = node.data.copy()

    def _get_start_node(self, mg, start, start_end_index):
        if start is None:
            not_aligned = mg.get_not_aligned_nodes()
            if len(not_aligned) == 0 :
                raise RuntimeError("cannot convert graph to tree no starting point")
            start = not_aligned[0]
            start_n = self._GraphtoTreeNode(None, None, start)
        else:
            start_n = self._GraphtoTreeNode(None, None, start)
            if start_end_index is not None:
                new_motif = self._reorient_motif(start_n.motif, start_end_index)
                start_n.motif = new_motif

        return start_n

    def _get_new_nodes(self, current):
        new_nodes = []
        for c in current.node.connections:
            if c is None:
                continue
            partner = c.partner(current.node.index)
            # check to see if motif is already in tree
            try:
                n = self.mt.get_node_by_id(partner.data.id)
            except:
                parent_end_index = c.end_index(current.node.index)
                node_end_index = c.end_index(partner.index)
                new_n = self._GraphtoTreeNode(current.node, parent_end_index,
                                              partner)
                new_n.motif = self._reorient_motif(new_n.motif, node_end_index)
                new_n.parent_end_index = self._get_new_parent_end_index(current.node, c)
                new_nodes.append(new_n)

        return new_nodes

    def _get_new_parent_end_index(self, parent, c):

        if parent.data.mtype != motif_type.HELIX:
            parent_end_index = c.end_index(parent.index)
            parent_end_name = parent.data.ends[parent_end_index].name()
            tree_parent = self.mt.get_node_by_id(parent.data.id)

            for i, end in enumerate(tree_parent.data.ends):
                if end.name() == parent_end_name:
                    return i

            raise RuntimeError("did not find original end something went really wrong")

        # helices always go end 0 to 1
        else:
            return 1

    def _reorient_motif(self, m, end_index):
        # nothing needs to be done
        if end_index == 0:
            return m

        if m.mtype != motif_type.HELIX:
            try:
                new_m = rm.manager.get_motif(name=m.name,
                                             end_name=m.ends[end_index].name())

            except exceptions.ResourceManagerException as e:
                raise exceptions.MotifTopologyException(
                    "cannot convert graph to tree because: " + e.message)

            new_m.copy_uuids_from_motif(m)
            return new_m

        else:
            if m.name[:5] == "HELIX":
                new_m = rm.manager.get_motif(name=m.name)
                new_m.copy_uuids_from_motif(m)
                return new_m

            else:
                spl = m.name.split("=")
                new_name = spl[1] + "=" + spl[0]
                new_m = rm.manager.get_motif(name=new_name)
                new_m.id = m.id
                return new_m

    def convert(self, mg, start=None, start_end_index=None, last_node=None):
        self.mt = motif_tree.MotifTree()
        self.mt.option('sterics', 0)

        start_n = self._get_start_node(mg, start, start_end_index)

        open_nodes = [ start_n ]
        last_node_to_add = None

        while len(open_nodes) > 0:
            current = open_nodes.pop(0)

            if last_node == current.node:
                last_node_to_add = current
                continue

            if current.parent is None:
                self.mt.add_motif(current.motif)
            else:
                new_parent_index = self.mt.get_node_by_id(
                                    current.parent.data.id).index
                self.mt.add_motif(current.motif, parent_index=new_parent_index,
                                  parent_end_index = current.parent_end_index)
            new_nodes = self._get_new_nodes(current)

            # check for duplicates
            for n1 in new_nodes:
                found = 0
                for n2 in open_nodes:
                    if n2.node.index == n1.node.index:
                        found = 1
                        break
                    if last_node_to_add is not None:
                        if last_node_to_add.node.index == n1.index:
                            found = 1
                            break

                if not found:
                    open_nodes.append(n1)


        if last_node_to_add is not None:
            new_parent_index = self.mt.get_node_by_id(
                last_node_to_add.parent.data.id).index
            self.mt.add_motif(last_node_to_add.motif,
                              parent_index=new_parent_index,
                              parent_end_index=last_node_to_add.parent_end_index)

        # add missing connections
        for n in mg:
            for c in n.connections:
                if c is None:
                    continue
                n_id = n.data.id
                partner = c.partner(n.index)
                pn_id = partner.data.id

                try:
                    mt_n1 = self.mt.get_node_by_id(n_id)
                    mt_n2 = self.mt.get_node_by_id(pn_id)
                except exceptions.MotifTreeException:
                    continue

                if mt_n1.parent == mt_n2:
                    continue
                if mt_n2.parent == mt_n1:
                    continue

                n1_ei = c.end_index(n.index)
                if n.data.mtype == motif_type.HELIX:
                    n1_ei = 1

                n1_end_name = n.data.ends[n1_ei].name()

                if self.mt.connections.in_connection(mt_n1.index, n1_end_name):
                    continue


                n2_ei = c.end_index(partner.index)
                if partner.data.mtype == motif_type.HELIX:
                    n2_ei = 1

                n2_end_name = partner.data.ends[n2_ei].name()
                self.mt.add_connection(mt_n1.index, mt_n2.index,
                                      n1_end_name, n2_end_name)



        self.mt.option('sterics', 1)
        return self.mt



def graph_to_tree(mg, start=None, last_end=None):
        if start is None:
            start = mg.graph.oldest_node()

        seen = []
        open = [ start ]

        mt = motif_tree.MotifTree()
        mt.option('sterics', 0)
        ss = mg.secondary_structure()
        seen_nodes = {}
        seen_connections = {}
        index = 0
        pos = 0

        while len(open) > 0:
            current = open.pop(0)
            ss_m = ss.motif(current.data.id)

            if index == 0:
                free_end = -1
                for i, c in enumerate(current.connections):
                    if c is None:
                        free_end = i

                name = ss_m.name
                if ss_m.name[2] == "=":
                   m =  rm.manager.get_motif(name = name)
                else:
                    m = rm.manager.get_motif(name = name,
                                            end_name=ss_m.ends[free_end].name())
                mt.add_motif(m)

            else:
                highest = -1
                p = None
                p_index = None
                p_end_index = None
                p_end_name = None
                c_end_index = -1
                c_end_name = None
                for c in current.connections:
                    if c is None:
                        continue
                    p_new = c.partner(current.index)
                    if p_new in seen_nodes:
                        p_index = seen_nodes[p_new]
                        if highest < p_index:
                            highest = p_index
                        else:
                            continue
                        p = p_new
                        p_end_index = c.end_index(p_new.index)
                        p_end_name = p.data.ends[p_end_index].name()
                        c_end_index = c.end_index(current.index)
                        c_end_name = current.data.ends[c_end_index].name()
                        if p.data.name[2] == "=":
                            act_parent = mt.get_node(p_index)
                            p_end_index = 1
                            p_end_name = act_parent.data.ends[1].name()

                if last_end is not None:
                    if last_end.name() == p_end_name and len(seen_nodes) != len(mg)-1:
                        open.append(current)
                        continue


                if ss_m.name[2] == "=":
                    name = ss_m.name
                    if c_end_index == 1:
                        spl = name.split("=")
                        name = spl[1] + "=" + spl[0]
                    m = rm.manager.get_motif(name = name)

                else:
                    m = rm.manager.get_motif(name = ss_m.name,
                                             end_name=c_end_name)

                #print m.name, p.data.name, p_index, p_end_index, p_end_name
                seen_connections[str(p_index) + " " + str(index)] = 1
                pos = mt.add_motif(m, parent_index=p_index, parent_end_name=p_end_name)
                if pos == -1:
                    raise ValueError("could not convert graph to tree")

            seen_nodes[current] = index
            index += 1
            for c in current.connections:
                if c is not None:
                    p = c.partner(current.index)
                    if p not in seen_nodes and p not in open:
                        open.append(p)

        for n in mg.graph:
            for c in n.connections:
                if c is None:
                    continue
                key1 = str(seen_nodes[c.node_1]) + " " + str(seen_nodes[c.node_2])
                key2 = str(seen_nodes[c.node_2]) + " " + str(seen_nodes[c.node_1])
                if key1 not in seen_connections and key2 not in seen_connections:
                    mt.add_connection(seen_nodes[c.node_1], seen_nodes[c.node_2])
                    seen_connections[key1] = 1

        return mt

File: src/motif_data_structure/motif_topology.cpp

//  motif_toplogy.cpp
//  RNAMake
//
//  Created by Joseph Yesselman on 1/8/16.
//  Copyright (c) 2016 Joseph Yesselman. All rights reserved.
//

#include <queue>
#include <map>

#include "motif_data_structure/motif_topology.h"
#include "resources/resource_manager.h"

namespace motif_data_structure {

MotifTreeOP
GraphtoTree::convert(
        MotifGraphOP const & mg,
        data_structure::graph::GraphNodeOP<motif::MotifOP> start,
        int start_end_index,
        data_structure::graph::GraphNodeOP<motif::MotifOP> last_node) {

    mt_ = std::make_shared<MotifTree>();
    mt_->set_option_value("sterics", false);

    auto start_node = _get_start_node(mg, start, start_end_index);
    auto last_node_to_add = _GraphtoTreeNodeOP(nullptr);
    auto current = _GraphtoTreeNodeOP(nullptr);
    auto open_nodes = std::queue<_GraphtoTreeNodeOP>();
    auto new_nodes = _GraphtoTreeNodeOPs();
    auto seen_nodes = std::map<int, int>();
    open_nodes.push(start_node);
    seen_nodes[start_node->node->index()] = 1;

    while (open_nodes.size() > 0) {
        current = open_nodes.front();
        open_nodes.pop();


        if (last_node == current->node) {
            last_node_to_add = current;
            //auto parent = mt_->get_node(current->parent->data()->id());
            continue;
        }

        int found = 0;
        try {
            mt_->get_node(current->motif->id());
            found = 1;
        } catch (MotifTreeException const & e) {}

        if (found) { continue; }


        if (current->parent == nullptr) {
            mt_->add_motif(current->motif);
        } else {
            auto parent = mt_->get_node(current->parent->data()->id());
            auto new_parent_index = parent->index();


            //currently a hack to make sure that if we start building not at an end and trying to add
            //somethign on the 0th end of a HELIX this stops it
            if (parent->data()->mtype() == util::MotifType::HELIX && parent->children()[1] != nullptr) {
                seen_nodes.erase(current->node->index());
                continue;
            }

            //can happen sometimes if you build from the middle of the graph
            if (current->parent_end_index == parent->data()->block_end_add()) {
                continue;
            }

            auto pos = mt_->add_motif(current->motif, new_parent_index, current->parent_end_index);
            if (pos == -1) {
                continue;
            }
        }

        new_nodes = _get_new_nodes(current);
        for (auto const & n : new_nodes) {
            //if(n == last_node_to_add) { continue; }
            int found = 0;
            try {
                mt_->get_node(n->motif->id());
                found = 1;
            } catch (MotifTreeException const & e) {}

            if (found) { continue; }

            if (seen_nodes.find(n->node->index()) == seen_nodes.end()) {
                open_nodes.push(n);
            }
        }

    }

    if (last_node_to_add != nullptr) {
        auto new_parent_index = mt_->get_node(last_node_to_add->parent->data()->id())->index();
        mt_->add_motif(last_node_to_add->motif,
                       new_parent_index,
                       last_node_to_add->parent_end_index);
    }

    // get all connections that need to added as extra connections to preserve topology
    for (auto const & n : *mg) {
        for (auto const c : n->connections()) {
            if (c == nullptr) { continue; }
            auto n_id = n->data()->id();
            auto partner = c->partner(n->index());
            auto pn_id = partner->data()->id();

            auto mt_n1 = data_structure::tree::TreeNodeOP<motif::MotifOP>(nullptr);
            auto mt_n2 = data_structure::tree::TreeNodeOP<motif::MotifOP>(nullptr);
            try {
                mt_n1 = mt_->get_node(n_id);
                mt_n2 = mt_->get_node(pn_id);
            }
            catch (MotifTreeException const & e) { continue; }

            if (mt_n1->parent() == mt_n2) { continue; }
            if (mt_n2->parent() == mt_n1) { continue; }

            auto n1_ei = c->end_index(n->index());
            if (n->data()->mtype() == util::MotifType::HELIX) { n1_ei = 1; }

            auto n1_end_name = n->data()->end_name(n1_ei);

            if (mt_->connections().in_connection(mt_n1->index(), n1_end_name)) { continue; }

            auto n2_ei = c->end_index(partner->index());
            if (partner->data()->mtype() == util::MotifType::HELIX) { n2_ei = 1; }

            auto n2_end_name = partner->data()->end_name(n2_ei);

            if (mt_->connections().in_connection(mt_n2->index(), n2_end_name)) { continue; }

            //std::cout << mt_n1->index() << " " << mt_n2->index() << " " << n1_end_name << " " << n2_end_name << std::endl;
            mt_->add_connection(mt_n1->index(), mt_n2->index(), n1_end_name, n2_end_name);

        }
    }


    mt_->set_option_value("sterics", false);
    return mt_;

}

GraphtoTree::_GraphtoTreeNodeOP
GraphtoTree::_get_start_node(
        MotifGraphOP const & mg,
        data_structure::graph::GraphNodeOP<motif::MotifOP> const & start,
        int start_end_index) {

    auto start_n = _GraphtoTreeNodeOP(nullptr);
    if (start == nullptr) {
        auto not_aligned = mg->unaligned_nodes();
        if (not_aligned.size() == 0) {
            throw MotifTopologyException("cannot convert graph to tree no starting point");
        }
        auto start = not_aligned[0];
        start_n = std::make_shared<_GraphtoTreeNode>(nullptr, 0, start, start->data());
    } else {
        start_n = std::make_shared<_GraphtoTreeNode>(nullptr, 0, start, start->data());
        start_n->motif = _get_reoriented_motif(start->data(), start_end_index);

    }

    return start_n;

}

motif::MotifOP
GraphtoTree::_get_reoriented_motif(
        motif::MotifOP const & m,
        int end_index) {

    // dont need to do anything already oriented
    if (end_index == 0) { return m; }

    if (m->mtype() != util::MotifType::HELIX) {
        try {
            auto new_m = resources::Manager::instance().motif(m->name(), "", m->end_name(end_index));
            new_m->copy_uuids_from_motif(*m);
            return new_m;
        }
        catch (resources::ResourceManagerException const & e) {
            throw MotifTopologyException(
                    String("cannot convert graph to tree because ") + e.what());
        }
    } else {
        if (m->name().substr(0, 5) == "HELIX") {
            auto new_m = resources::Manager::instance().motif(m->name());
            new_m->copy_uuids_from_motif(*m);
            return new_m;


        } else {
            auto spl = base::split_str_by_delimiter(m->name(), "=");
            auto new_name = spl[1] + "=" + spl[0];
            auto new_m = resources::Manager::instance().motif(m->name());
            //since each basepair step has a different motif to represent it
            //these are likly two different motifs cannot do copy_uuids_from_motif
            new_m->id(m->id());
            return new_m;
        }

    }

}


GraphtoTree::_GraphtoTreeNodeOPs
GraphtoTree::_get_new_nodes(
        GraphtoTree::_GraphtoTreeNodeOP const & current) {

    auto new_nodes = _GraphtoTreeNodeOPs();
    auto partner = data_structure::graph::GraphNodeOP<motif::MotifOP>();
    auto n = data_structure::tree::TreeNodeOP<motif::MotifOP>();
    auto found = 1;

    for (auto const & c : current->node->connections()) {
        if (c == nullptr) { continue; }

        found = 1;
        partner = c->partner(current->node->index());
        // check to see if motif is already in tree
        try {
            n = mt_->get_node(partner->data()->id());
        } catch (...) { found = 0; }


        if (found) { continue; }

        auto parent_end_index = c->end_index(current->node->index());
        auto node_end_index = c->end_index(partner->index());
        auto new_n = std::make_shared<_GraphtoTreeNode>(current->node, parent_end_index,
                                                        partner, partner->data());
        new_n->motif = _get_reoriented_motif(new_n->motif, node_end_index);
        new_n->parent_end_index = _get_new_parent_end_index(current->node, c);

        //if(new_n->parent_end_index != 0) { continue; }

        new_nodes.push_back(new_n);

    }

    return new_nodes;

}

int
GraphtoTree::_get_new_parent_end_index(
        data_structure::graph::GraphNodeOP<motif::MotifOP> const & parent,
        data_structure::graph::GraphConnectionOP<motif::MotifOP> const & c) {

    if (parent->data()->mtype() != util::MotifType::HELIX) {
        auto parent_end_index = c->end_index(parent->index());
        auto parent_end_name = parent->data()->end_name(parent_end_index);
        auto tree_parent = mt_->get_node(parent->data()->id());

        int i = -1;
        for (auto const & end : tree_parent->data()->ends()) {
            i++;
            if (end->get_name_str() == parent_end_name) {
                return i;
            }
        }

        throw std::runtime_error("did not find original end something went really wrong");
    }

        //helices always go end 0 to 1
    else {
        return 1;
    }

}


MotifTreeOP
graph_to_tree(
        MotifGraphOP const & mg,
        data_structure::graph::GraphNodeOP<motif::MotifOP> start,
        structure::BasepairOP last_end) {

    if (start == nullptr) {
        start = mg->oldest_node();
    }

    auto mt = std::make_shared<MotifTree>();
    mt->set_option_value("sterics", false);
    auto open = std::queue<data_structure::graph::GraphNodeOP<motif::MotifOP>>
            ();
    auto current = data_structure::graph::GraphNodeOP<motif::MotifOP>(nullptr);
    auto index = 0;
    auto m = motif::MotifOP(nullptr);
    open.push(start);

    auto seen_nodes = std::map<data_structure::graph::GraphNodeOP<motif::MotifOP>,
            int>();
    auto current_nodes = std::map<data_structure::graph::GraphNodeOP<motif::MotifOP>,
            int>();
    auto seen_connections = std::map<String, int>();
    current_nodes[start] = 1;

    int highest = -1, pos = -1;
    auto p = data_structure::graph::GraphNodeOP<motif::MotifOP>(nullptr);
    auto p_new = data_structure::graph::GraphNodeOP<motif::MotifOP>(nullptr);
    auto p_end_name = String("");
    auto p_end_index = -1;
    auto p_index = -1;
    auto c_end_index = -1;
    auto c_end_name = String("");

    while (open.size() > 0) {
        current = open.front();
        open.pop();

        if (index == 0) {
            int free_end = -1;
            int i = 0;
            for (auto const & c : current->connections()) {
                if (c == nullptr) {
                    free_end = i;
                    break;
                }
            }
            assert(free_end != -1 && "no free end, cannot convert to motif tree");

            auto name = current->data()->name();
            if (name[2] == '=') {
                if (free_end == 1) {
                    auto spl = base::split_str_by_delimiter(name, "=");
                    name = spl[1] + "=" + spl[0];
                }

                m = resources::Manager::instance().motif(name);
            } else {
                m = resources::Manager::instance().motif(name, "", current->data()->ends()[0]->get_name_str());
            }
            mt->add_motif(m);
            mt->get_node(0)->data()->id(current->data()->id());
        } else {

            highest = -1;
            p_end_index = -1;
            c_end_index = -1;
            p_end_name = "";
            c_end_name = "";
            p = nullptr;

            for (auto const & c : current->connections()) {
                if (c == nullptr) { continue; }
                p_new = c->partner(current->index());
                if (seen_nodes.find(p_new) == seen_nodes.end()) { continue; }
                p_index = seen_nodes[p_new];
                if (highest > p_index) { continue; }
                highest = p_index;
                p = p_new;
                p_end_index = c->end_index(p->index());
                p_end_name = p->data()->ends()[p_end_index]->get_name_str();
                c_end_index = c->end_index(current->index());
                c_end_name = current->data()->ends()[c_end_index]->get_name_str();

            }

            if (last_end != nullptr) {
                if (last_end->get_name_str() == p_end_name &&
                    seen_nodes.size() != mg->size() - 1) {
                    open.push(current);
                    continue;
                }
            }


            assert(p != nullptr && "could not find parent something really wrong");
            //std::cout << p->data()->name() << " " << p_end_index << std::endl;
            if (p->data()->name()[2] == '=') {
                p_end_index = 1;
                auto act_parent = mt->get_node(seen_nodes[p]);
                p_end_name = act_parent->data()->ends()[1]->get_name_str();
            }

            if (current->data()->name()[2] == '=' && c_end_index == 1) {
                auto spl = base::split_str_by_delimiter(current->data()->name(), "=");
                auto name = spl[1] + "=" + spl[0];
                m = resources::Manager::instance().motif(name);
            } else {
                m = resources::Manager::instance().motif(current->data()->name(), "", c_end_name);
            }

            seen_connections[std::to_string(p_index) + " " + std::to_string(index)] = 1;
            pos = mt->add_motif(m, p_index, p_end_name);
            mt->get_node(pos)->data()->id(current->data()->id());
            assert(pos != -1 && "did not sucessfully add motif to tree during conversion");

        }

        seen_nodes[current] = index;
        current_nodes.erase(current);
        index++;

        for (auto const & c : current->connections()) {
            if (c == nullptr) { continue; }
            p = c->partner(current->index());
            if (seen_nodes.find(p) != seen_nodes.end()) { continue; }
            if (current_nodes.find(p) != current_nodes.end()) { continue; }
            open.push(p);
            current_nodes[p] = 1;
        }
    }

    String key1, key2;
    for (auto const & n : *mg) {
        for (auto const & c : n->connections()) {
            if (c == nullptr) { continue; }
            key1 = std::to_string(seen_nodes[c->node_1()]) + " " + std::to_string(seen_nodes[c->node_2()]);
            key2 = std::to_string(seen_nodes[c->node_2()]) + " " + std::to_string(seen_nodes[c->node_1()]);
            if (seen_connections.find(key1) == seen_connections.end() &&
                seen_connections.find(key2) == seen_connections.end()) {
                seen_connections[key1] = 1;
                mt->add_connection(seen_nodes[c->node_1()], seen_nodes[c->node_2()],
                                   "", "");
            }
        }
    }


    return mt;

}

}



















